### PR TITLE
Add option to avoid cuda init to FakeTensorMode

### LINKED ISSUE #122109: Tensor creation under FakeMode requires real device
### 🚀 The feature, motivation and pitch

On a CPU-only machine, creating a tensor on a fake cuda device invokes CUDA device initialization: “RuntimeError: Found no NVIDIA driver on your system.” 

```
import torch
from torch._subclasses.fake_tensor import FakeTensorMode

torch.cuda.is_available()  # returns False
device = torch.device("cuda:0")

with FakeTensorMode():
   t1 = torch.empty(10, device=device)
   t2 = torch.ones(10, device=device)
   t3 = torch.zeros(10, device=device)
   t4 = torch.rand(10, device=device)
   t5 = torch.tensor([1,2,3], device=device)
```

Fake tensor creation shouldn’t require real physical device. 

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @bdhirsh @anijain2305 @chauhang @eellison

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122668

Also partially fixes #122109

This PR:
- adds an option to avoid cuda init to FakeTensorMode (default: False).
  This is an option because changing this is BC-breaking and has perf
  implications.
- Concretely, the option (a) prevents constant propagation of ops that
  move the tensor to a new device and (b) changes the torch.tensor ctor
  strategy.
- We add a C++ flag (lift_then_h2d) to toggle the torch.tensor(1, device='cuda') ctor
  strategy. When false (default), it does the current PyTorch behavior
  of unconditionally constructing a concrete CUDA tensor then calling
  lift_fresh on it. When true, we instead construct a concrete CPU
  tensor, call lift_fresh, and then call Tensor.to (under any ambient
  modes).

Test Plan:
- existing tests

Differential Revision: [D55591785](https://our.internmc.facebook.com/intern/diff/D55591785)

### CODE REVIEW

**Comment by ezyang** [comment]
@eellison constant prop is always on tensors that are not too big, right? I suggest we just always do constant prop on CPU.

**Comment by eellison** [comment]
@ezyang right it is always on tensor.numel() == 1

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -869,23 +877,30 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
...
     def __enter__(self):
+        prev_lift_then_h2d = None
+        if self.avoid_device_init:
+            # See NOTE: [torch.tensor, lift_fresh, and device movement]
+            prev_lift_then_h2d = torch._C._get_lift_then_h2d()
+            torch._C._set_lift_then_h2d(True)
         maybe_prev_fake_mode = torch._C._unset_dispatch_mode(self._mode_key)
         if self is not maybe_prev_fake_mode:
-            self.enter_stack.append((True, maybe_prev_fake_mode))
+            self.enter_stack.append((True, maybe_prev_fake_mode, prev_lift_then_h2d))
```
Comment (bdhirsh) [inline torch/_subclasses/fake_tensor.py:887]: Storing the config directly on FakeTensorMode does give us an annoying amount of flexibility when handling nested modes.

WDYT of just having the config be global state? Downside: you get weird behavior if you change the global state mid-way through the usage of your FakeTensorMode. Maybe that's a reason to stick with setting the config on FakeTensorMode directly, although the expected usage of the config is probably to globally enable it during all of export / model init.

File: torch/csrc/utils/tensor_new.cpp [hunk 2]
```diff
@@ -453,19 +455,32 @@ Tensor internal_new_from_data(
     // "no observable data dependence".  In an ideal world, we wouldn't trace
     // a to() call but I need to think harder about what exactly we should trace
     // in this case.
-    tensor = tensor.to(
-        device, inferred_scalar_type, /*non_blocking=*/false, /*copy=*/false);
+    if (!get_lift_then_h2d()) {
+      tensor = tensor.to(
+          device, inferred_scalar_type, /*non_blocking=*/false, /*copy=*/false);
```
Comment (bdhirsh) [inline torch/csrc/utils/tensor_new.cpp:460]: Similar to my comment in `fake_tensor.py`, but: do we care about const-prop of dtype conversions still happening when this config is on? IIUC, we will still want to unconditionally do a `tensor.to(inferred_scalar_type)` to get that to work

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (zou3519) [reply torch/_subclasses/fake_tensor.py:887]: What's the problem with nested modes? We set and reset the state correctly on `__enter__/__exit__`.

File: torch/_subclasses/fake_tensor.py [hunk 3]
```diff
@@ -1282,6 +1297,15 @@ def maybe_to_constant(t):
             if type(args[0]) is torch.Tensor:
                 return converter.from_real_tensor(self, args[0])
 
+        # If we are trying to avoid device init, then we need to avoid constant
+        # prop on constant tensors for ops that change devices.
+        avoiding_device_init = False
+        if self.avoid_device_init:
+            if func == torch.ops.aten._to_copy.default and "device" in kwargs:
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_subclasses/fake_tensor.py:1304]: yup

File: torch/csrc/utils/tensor_new.cpp (hunk 2 above)
Comment (zou3519) [reply torch/csrc/utils/tensor_new.cpp:460]: hmm, the config is supposed to change the ordering of `lift_fresh` w.r.t. to the tensor.to() call. Having something more fine-grained ("only do this for devices") adds more complexity to the code then I'd like ("lift then .to()" vs ".to() then lift" is easier to reason about).

If this is really a problem then we can make it more fine-grained in the future. The current solution doesn't prevent us from doing that.

I'll change the name of the config from `lift_then_h2d` to `lift_then_to` to better reflect the semantics

File: torch/_subclasses/fake_tensor.py [hunk 4]
```diff
@@ -791,6 +791,7 @@ def __init__(
         allow_non_fake_inputs=False,
         shape_env=None,
         static_shapes=None,
+        avoid_device_init=False,
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:794]: Do we need this as a flag ? Could we just set it unconditionally if we're trying to run cuda fake tensors with no cuda device available ?

**Review by eellison — COMMENTED** [review]
Would be great to add support for constant prop, maybe as a separate pr on this stack.

File: torch/csrc/utils/tensor_new.cpp (hunk 2 above)
Comment (bdhirsh) [reply torch/csrc/utils/tensor_new.cpp:460]: sgtm (agreed it is extra complexity, and I don't have any real signal on how important const-propping dtype conversions is)

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (bdhirsh) [reply torch/_subclasses/fake_tensor.py:887]: I think I'm mostly just commenting on the fact that it feels unusual to do something like this:
```
with FakeTensorMode(prev_lift_then_h2d=True):
    foo()
    with FakeTensorMode(prev_lift_then_h2d=False):
        bar()
```

But we are giving users the full flexibility to let people do this, and we need to handle it via the extra state reseting on enter/exit.

I agree that this extra state resetting isn't actually that complicated. So if you want to keep the state on the mode then sgtm, I just figured I would call it out.

**Review by bdhirsh — APPROVED** [review]
sgtm!

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (eellison) [reply torch/_subclasses/fake_tensor.py:887]: See my other comment. Can we avoid adding this as a ctor and detect when it is necessary if cuda device is used and device is not avialable ?

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55591785).

File: torch/_subclasses/fake_tensor.py [hunk 5]
```diff
@@ -761,6 +761,7 @@ def __init__(
         allow_non_fake_inputs=False,
         shape_env=None,
         static_shapes=None,
+        _allow_unsafe_data_ptr_access=True,
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:764]: Can we avoid adding this ctor option if this is not something we want around permanently ? Maybe take it from a config from functorch ?

Comment (zou3519) [reply torch/_subclasses/fake_tensor.py:764]: Yup

**Review by SherlockNoMad — APPROVED** [review]

**Comment by SherlockNoMad** [comment]
gentle ping for landing this.

**Comment by zou3519** [comment]
Summary of changes:
- we now auto-detect if we need to use the new option
- the new option is "construct the tensor on cpu, call lift_fresh, and do a final .to(device) later" instead of "construct the tensor on cuda and then call lift_fresh"

**Review by eellison — APPROVED** [review]

**Comment by SherlockNoMad** [comment]
land this via fbcode?

**Comment by zou3519** [comment]
This was landed in https://github.com/pytorch/pytorch/pull/124413

**Pull request closed without merging**